### PR TITLE
feat(application-system): Official journal of Iceland application

### PR DESCRIPTION
**Pull request by jonbjarnio**
## What

Creating a new application and client to communicate with DMR.

## Why

This application purpose is to make life easier for people who work with the law, by allowing the to submit changes to regulations and such through the application system.

image
image
image

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/application/templates/official-journal-of-iceland/src/graphql/queries.ts
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/graphql/queries.ts:1]: Can this be a screen specific gql query?

File: libs/application/templates/official-journal-of-iceland/src/fields/NewCase/NewCase.tsx [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                    ...newCase,
+                    signatureType:
+                      value?.value as MinistryOfJusticeCase['signatureType'],
+                  })
+                }}
+              />
+            </Box>
+            <Box>
+              <Button
+                onClick={() => console.log('fetch old signatures')}
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/NewCase/NewCase.tsx:243]: :eyes:

File: libs/application/templates/official-journal-of-iceland/src/fields/NewCase/TemplateModal.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            q: lazyFilter,
+          },
+        },
+        fetchPolicy: 'no-cache',
+        onCompleted(data) {
+          setLoading(false)
+          setTemplates(data.ministryOfJusticeSearchCaseTemplates.items)
+        },
+        onError(error) {
+          console.log(error)
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/NewCase/TemplateModal.tsx:66]: Should this be handled more?

File: libs/application/templates/official-journal-of-iceland/src/fields/OriginalData/OriginalData.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import { FormGroup } from '../../components/FromGroup/FormGroup'
+import { useFormatMessage } from '../../hooks'
+import { FILE_SIZE_LIMIT, UPLOAD_ACCEPT } from '../../lib/constants'
+import { originalData } from '../../lib/messages'
+import { OJOIFieldBaseProps } from '../../lib/types'
+
+export const OriginalData = ({ application }: OJOIFieldBaseProps) => {
+  const { f } = useFormatMessage(application)
+
+  // TODO: handle file upload
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/OriginalData/OriginalData.tsx:12]: :eyes:

File: libs/application/templates/official-journal-of-iceland/src/fields/Prerequisites/Prerequisites.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import { FormWrap } from '../../components/FormWrap/FormWrap'
+import { useFormatMessage } from '../../hooks'
+import { general, prerequisites } from '../../lib/messages'
+import { OJOIFieldBaseProps } from '../../lib/types'
+
+export const Prerequisites = ({ application, refetch }: OJOIFieldBaseProps) => {
+  const { f, locale } = useFormatMessage(application)
+  const [updateApplication, { loading }] = useMutation(UPDATE_APPLICATION)
+  const [submitApplication] = useMutation(SUBMIT_APPLICATION, {
+    onError: (e) => console.error(e.message),
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/Prerequisites/Prerequisites.tsx:19]: Should this be handled more?

File: libs/application/templates/official-journal-of-iceland/src/fields/Preview/Preview.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import { FormIntro } from '../../components/FormIntro/FormIntro'
+import { useFormatMessage } from '../../hooks'
+import { preview } from '../../lib/messages'
+import { OJOIFieldBaseProps } from '../../lib/types'
+
+export const Preview = ({ application }: OJOIFieldBaseProps) => {
+  const { f } = useFormatMessage(application)
+
+  const copyPreviewUrl = () => {
+    console.log('replace me with logic')
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/Preview/Preview.tsx:11]: :eyes:

File: libs/application/templates/official-journal-of-iceland/src/fields/Preview/Preview.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  const fetchPdf = () => {
+    console.log('replace me with logic')
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/Preview/Preview.tsx:15]: :eyes:

File: libs/application/templates/official-journal-of-iceland/src/fields/Preview/Preview.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
...DAwMDMxIDY1NTM1IGYNCjAwMDAwMDAwMzIgNjU1MzUgZg0KMDAwMDAwMDAzMyA2NTUzNSBmDQowMDAwMDAwMDM0IDY1NTM1IGYNCjAwMDAwMDAwMzUgNjU1MzUgZg0KMDAwMDAwMDAzNiA2NTUzNSBmDQowMDAwMDAwMDM3IDY1NTM1IGYNCjAwMDAwMDAwMzggNjU1MzUgZg0KMDAwMDAwMDAzOSA2NTUzNSBmDQowMDAwMDAwMDQwIDY1NTM1IGYNCjAwMDAwMDAwNDEgNjU1MzUgZg0KMDAwMDAwMDA0MiA2NTUzNSBmDQowMDAwMDAwMDQzIDY1NTM1IGYNCjAwMDAwMDAwNDQgNjU1MzUgZg0KMDAwMDAwMDA0NSA2NTUzNSBmDQowMDAwMDAwMDQ2IDY1NTM1IGYNCjAwMDAwMDAwNDcgNjU1MzUgZg0KMDAwMDAwMDA0OCA2NTUzNSBmDQowMDAwMDAwMDQ5IDY1NTM1IGYNCjAwMDAwMDAwNTAgNjU1MzUgZg0KMDAwMDAwMDA1MSA2NTUzNSBmDQowMDAwMDAwMDUyIDY1NTM1IGYNCjAwMDAwMDAwNTMgNjU1MzUgZg0KMDAwMDAwMDA1NCA2NTUzNSBmDQowMDAwMDAwMDAwIDY1NTM1IGYNCjAwMDAwMDc0NTcgMDAwMDAgbg0KMDAwMDAwNzczNiAwMDAwMCBuDQowMDAwMDEwOTE2IDAwMDAwIG4NCjAwMDAwMTA5NjEgMDAwMDAgbg0KdHJhaWxlcg0KPDwvU2l6ZSA1OS9Sb290IDEgMCBSL0luZm8gMTMgMCBSL0lEWzxGN0M0MjU5ODA4MEUwRDQ0ODNFQjg5OTU0MkVBMjZBOT48RjdDNDI1OTgwODBFMEQ0NDgzRUI4OTk1NDJFQTI2QTk+XSA+Pg0Kc3RhcnR4cmVmDQoxMTMxOA0KJSVFT0YNCnhyZWYNCjAgMA0KdHJhaWxlcg0KPDwvU2l6ZSA1OS9Sb290IDEgMCBSL0luZm8gMTMgMCBSL0lEWzxGN0M0MjU5ODA4MEUwRDQ0ODNFQjg5OTU0MkVBMjZBOT48RjdDNDI1OTgwODBFMEQ0NDgzRUI4OTk1NDJFQTI2QTk+XSAvUHJldiAxMTMxOC9YUmVmU3RtIDEwOTYxPj4NCnN0YXJ0eHJlZg0KMTI2NTUNCiUlRU9G"
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/Preview/Preview.tsx:52]: whoah

File: libs/application/templates/official-journal-of-iceland/src/fields/PublishingPrefrences/PublishingPrefrences.tsx [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+import { useFormatMessage } from '../../hooks'
+import { error, general, publishingPrefrences } from '../../lib/messages'
+import { BooleanValue, OJOIFieldBaseProps } from '../../lib/types'
+import { getWeekdayDates, isValidEmail, isValidPhone } from '../../lib/utils'
+export type CommunicationChannel = {
+  phone: string
+  email: string
+}
+
+export const PublishingPrefrences = ({ application }: OJOIFieldBaseProps) => {
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/PublishingPrefrences/PublishingPrefrences.tsx:22]:
```suggestion
export const PublishingPreferences = ({ application }: OJOIFieldBaseProps) => {
```

File: libs/application/templates/official-journal-of-iceland/src/fields/Summary/Summary.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  type: 'Auglýsing',
+  caseTitle:
+    'AUGLÝSING um skrá yfir þau störf hjá Múlaþingi sem eru undanskilin verkfallsheimild.',
+  caseDepartment: 'B-Deild',
+  dateOfSubmission: '25.07.2023',
+  caseFastTrack: 'Nei',
+  estimatedDateOfPublication: '25.10.2023',
+  estimatedPrice: '23.000',
+  caseClassicifation: 'Verkföll og vinnudeilur',
+}
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/fields/Summary/Summary.tsx:20]: Should probably be in a separate file

File: libs/application/templates/official-journal-of-iceland/src/forms/Complete.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  newCase,
+  originalData,
+  prerequisites,
+  preview,
+  publishingPrefrences,
+  summary,
+} from '../lib/messages'
+export const Complete: Form = buildForm({
+  id: 'OfficalJournalOfIcelandPreRequsitesForm',
+  title: 'Skilyrði',
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/forms/Complete.ts:19]: contentful?

File: libs/application/templates/official-journal-of-iceland/src/lib/OJOIApplication.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    ApplicationConfigurations.OfficialJournalOfIceland.translation,
+  ],
+  dataSchema: OfficalJournalOfIcelandSchema,
+  allowMultipleApplicationsInDraft: true,
+  stateMachineConfig: {
+    initial: ApplicationStates.PREREQUISITS,
+    states: {
+      [ApplicationStates.PREREQUISITS]: {
+        meta: {
+          name: 'Umsókn um stjórnartíðindi',
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/lib/OJOIApplication.ts:59]: translate?

File: libs/application/templates/official-journal-of-iceland/src/lib/constants.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+export const emailRegex = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i
+
+export const UPLOAD_ACCEPT = '.pdf'
+
+export const FILE_SIZE_LIMIT = 10000000
+
+export const VERDSKRA_LINK =
+  'https://www.stjornartidindi.is/PdfVersions.aspx?recordId=0f574646-eb9d-430b-bbe7-936e7c9389a0'
```
Comment (thorkellmani) [inline libs/application/templates/official-journal-of-iceland/src/lib/constants.ts:8]: Will this never change?

**Review by thorkellmani — COMMENTED** [review]
Just a few things i noticed

File: apps/application-system/api/infra/application-system-api.ts [hunk 13]
```diff
@@ -270,6 +263,7 @@ export const serviceSetup = (services: {
       DirectorateOfImmigration,
       SocialInsuranceAdministration,
       OccupationalLicenses,
+      WorkMachines,
```
Comment (saevarma) [inline apps/application-system/api/infra/application-system-api.ts:266]: ~This seems unrelated to the topic of the PR?

File: libs/api/domains/ministry-of-justice/src/lib/ministryOfJustice.resolver.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+} from '@island.is/auth-nest-tools'
+import { ApiScope } from '@island.is/auth/scopes'
+import { Audit } from '@island.is/nest/audit'
+import { UseGuards } from '@nestjs/common'
+import { Args, Query, Resolver } from '@nestjs/graphql'
+import { MinistryOfJusticeService } from './ministryOfJustice.service'
+import { SearchCaseTemplateInput } from './models/SearchCaseTemplate.input'
+import { SearchCaseTemplateResponse } from './models/SearchCaseTemplate.response'
+
+@UseGuards(IdsUserGuard, IdsAuthGuard)
```
Comment (saevarma) [inline libs/api/domains/ministry-of-justice/src/lib/ministryOfJustice.resolver.ts:16]: You don't need both the guards. `IdsUserGuard` suffice, and that guard forces the `nationalId` claim to be present in the access token, to force end user authentication. While `IdsAuthGuard` requires a valid access token where the `nationalId` claim is optional, so it allows machine client tokens, which we don't want here.

File: libs/api/domains/ministry-of-justice/src/lib/ministryOfJustice.resolver.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+  @Query(() => SearchCaseTemplateResponse, {
+    name: 'ministryOfJusticeSearchCaseTemplates',
+  })
+  @Audit()
+  async searchCaseTemplates(
+    @CurrentUser() user: User,
+    @Args('input') input: SearchCaseTemplateInput,
+  ) {
+    return await this.ministryOfJusticeService.searchCaseTemplates(user, input)
```
Comment (saevarma) [inline libs/api/domains/ministry-of-justice/src/lib/ministryOfJustice.resolver.ts:32]: ~ Redundant `await`

File: libs/api/domains/ministry-of-justice/src/lib/models/CaseDepartment.model.ts
Comment (saevarma) [inline libs/api/domains/ministry-of-justice/src/lib/models/CaseDepartment.model.ts:1]: ~ Inconsistency in `CaseCategory`, `CaseDepartment`, `CaseSignature`, `CaseSubCategory` and `CaseTemplate` file name. We don't have a strict code standard for this but at least keep it consistent in the scope of the library. The code itself is consistent so my 2cent would be to rename the files to `caseCategory.type.ts`, `caseDepartment.type.ts`, `caseSignature.type.ts`, `caseSubCategory.type.ts` and `caseTemplate.type.ts`.

File: libs/api/domains/ministry-of-justice/src/lib/models/SearchCaseTemplate.response.ts [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import { Field, Int, ObjectType } from '@nestjs/graphql'
+import { Case } from './Case.model'
+
+@ObjectType('MinistryOfJusticeSearchCaseTemplateResponse')
+export class SearchCaseTemplateResponse {
```
Comment (saevarma) [inline libs/api/domains/ministry-of-justice/src/lib/models/SearchCaseTemplate.response.ts:5]: You can use the `PagintedResponse` function from the `@island.is/nest/pagination` library to extend from which add the `totalCount` and `pageInfo` props.

File: apps/application-system/api/infra/application-system-api.ts (hunk 13 above)
Comment (jonbjarnio) [reply apps/application-system/api/infra/application-system-api.ts:269]: I think somebody else forgot to add this with their changes, I had to add this in order for the feature deploymenet to successfully deploy.

File: libs/api/domains/ministry-of-justice/src/lib/ministryOfJustice.service.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
+import type { User } from '@island.is/auth-nest-tools'
+import type { Logger } from '@island.is/logging'
+import { LOGGER_PROVIDER } from '@island.is/logging'
+import { isDefined } from '@island.is/shared/utils'
+import { Inject, Injectable } from '@nestjs/common'
+import { CaseSignatureType } from './models/caseSignature.type'
+import { SearchCaseTemplateInput } from './models/searchCaseTemplate.input'
+import { PaginatedSearchCaseTemplateResponse } from './models/searchCaseTemplate.response'
+import { Case } from './models/case.model'
```
Comment (saevarma) [inline libs/api/domains/ministry-of-justice/src/lib/ministryOfJustice.service.ts:9]: Unused imports?

**Review by saevarma — APPROVED** [review]
LGTM

File: apps/application-system/api/infra/application-system-api.ts [hunk 18]
```diff
@@ -63,7 +55,7 @@ export const workerSetup =
           staging: 'https://identity-server.staging01.devland.is',
           prod: 'https://innskra.island.is',
         },
-        XROAD_CHARGE_FJS_V2_PATH: {
+        CHARGE_FJS_V2_PATH: {
```
Comment (busla) [inline apps/application-system/api/infra/application-system-api.ts:58]: perhaps use `XROAD_` as an env prefix for xroad vars?

**Review by busla — CHANGES_REQUESTED** [review]
LGTM, minor concern

File: libs/shared/form-fields/src/lib/CheckboxController/CheckboxController.tsx [hunk 19]
```diff
@@ -53,7 +53,10 @@ export const CheckboxController: FC<
 }) => {
   const { clearErrors, setValue } = useFormContext()
 
+  // eslint-disable-next-line func-style
   function handleSelect(option: Option, checkedValues: string[]) {
+    console.log('option', option)
+    console.log('checkedValues', checkedValues)
```
Comment (Tryggvig) [inline libs/shared/form-fields/src/lib/CheckboxController/CheckboxController.tsx:59]:
```suggestion
    
```

**Review by Tryggvig — APPROVED** [review]
Approved with comments

**Review by busla — APPROVED** [review]
LGTM

File: libs/application/core/src/lib/fieldBuilders.ts [hunk 20]
```diff
@@ -43,7 +43,11 @@ import {
 } from '@island.is/application/types'
 
 import { Colors } from '@island.is/island-ui/theme'
-import { SpanType } from '@island.is/island-ui/core/types'
+import {
+  ResponsiveProp,
```
Comment (Thorrik58) [inline libs/application/core/src/lib/fieldBuilders.ts:47]: remove?

**Review by Thorrik58 — APPROVED** [review]
LGTM
